bokeh/client/session.py

```diff
@@ -153,11 +153,16 @@ def __init__(self, session_id=DEFAULT_SESSION_ID, io_loop=None, url=DEFAULT_SERV
         self._connection = ClientConnection(session=self, io_loop=io_loop, url=url)
 
         self._current_patch = None
+        self._callbacks = []
 
     def _attach_document(self, document):
         self._document = document
         self._document.on_change(self._document_changed)
 
+    def add_periodic_callback(self, cb, callback_time):
```

Comment:
I would make periodic its own thing, because it needs an event (like the RootAddedEvent/RootRemovedEvent) allowing the session to track them. on change is special because it's how you get events. 

to be clear the way I see it working is that Document has no knowledge of ioloop it only has the callback info, session has to add/remove from the loop when callbacks are added and removed.

good point we might want one shot timeouts in here too. That could be unified with periodic I think and the session would look at callback type and decide how to add it to the ioloop.